internal/controller/server/executor_services.go

```diff
@@ -59,3 +59,39 @@ func (e *executorCPServicesServer) Register(ctx context.Context, request *execut
 	}
 	return res, err
 }
+
+func (e *executorCPServicesServer) FetchJob(ctx context.Context, executorData *executorCPproto.ExecutorID) (*executorCPproto.Job, error) {
+	uuid, err := e.idgen.Generate()
+	if err != nil {
+		log.Error(err, fmt.Sprintf("executor id: %s, error while assigning id to the request", executorData.Id))
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	ctx = context.WithValue(ctx, util.ContextKeyUUID, uuid)
+
+	res, err := e.procExec.GetJob(ctx, executorData)
+	if err != nil {
+		if err.Error() == status.Error(codes.NotFound, constant.Controller+"no pending job").Error() {
+			return &executorCPproto.Job{HasJob: "no"}, nil
+		}
+		log.Error(err, fmt.Sprintf("request id: %v, executor id: %s, error while assigning job to executor", uuid, executorData.Id))
+		return nil, err
+	}
+	return res, err
+}
+
+func (e *executorCPServicesServer) SendExecutionContext(ctx context.Context, executionData *executorCPproto.ExecutionContext) (*executorCPproto.Acknowledgement, error) {
+	uuid, err := e.idgen.Generate()
+	if err != nil {
+		log.Error(err, fmt.Sprintf("executor id: %s, error while assigning id to the request", executionData.ExecutorID))
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	ctx = context.WithValue(ctx, util.ContextKeyUUID, uuid)
+	log.Info(fmt.Sprintf("request id: %v, recieved execution data: %+v", uuid, executionData))
+	err = e.procExec.SaveJobExecutionData(ctx, executionData)
+	if err != nil {
+		log.Error(err, fmt.Sprintf("request id: %v, executor id: %s, error while saving job execution logs", uuid, executionData.ExecutorID))
+		return &executorCPproto.Acknowledgement{Recieved: true}, err
+	}
+	return &executorCPproto.Acknowledgement{Recieved: true}, nil
```

Comment:
You can replace nil with err in 96th line and remove 94th line

Reply:
this is not a best practice. As such we have to check error and correspondingly return it